worker/src/ocall_bridge/attestation_ocall_impl.rs

```diff
@@ -0,0 +1,180 @@
+        report: sgx_report_t,
+        quote_type: sgx_quote_sign_type_t,
+        spid: sgx_spid_t,
+        quote_nonce: sgx_quote_nonce_t,
+    ) -> OCallBridgeResult<(sgx_report_t, Vec<u8>)> {
+        let mut real_quote_len: u32 = 0;
+
+        let (p_sig_rl, sig_rl_size) = vec_to_c_pointer_with_len(revocation_list);
+
+        let ret =
+            unsafe { sgx_calc_quote_size(p_sig_rl, sig_rl_size, &mut real_quote_len as *mut u32) };
+
+        ensure!(
+            ret == sgx_status_t::SGX_SUCCESS,
+            OCallBridgeError::GetQuote(ret)
+        );
+
+        debug!("    Quote size = {}", real_quote_len);
+
+        let p_report = &report as *const sgx_report_t;
+        let p_spid = &spid as *const sgx_spid_t;
+        let p_nonce = &quote_nonce as *const sgx_quote_nonce_t;
+
+        let mut qe_report = sgx_report_t::default();
+        let p_qe_report = &mut qe_report as *mut sgx_report_t;
+
+        const RET_QUOTE_BUF_LEN: usize = 2048;
+        let mut return_quote_buf = [0u8; RET_QUOTE_BUF_LEN];
+        let p_quote = return_quote_buf.as_mut_ptr();
+
+        if real_quote_len > RET_QUOTE_BUF_LEN as u32 {
+            error!(
+                "   effective quote length ({}) exceeds buffer size ({})",
+                real_quote_len, RET_QUOTE_BUF_LEN
+            );
+            return Err(OCallBridgeError::GetQuote(
+                sgx_status_t::SGX_ERROR_FAAS_BUFFER_TOO_SHORT,
+            ));
+        }
+
+        let ret = unsafe {
+            sgx_get_quote(
+                p_report,
+                quote_type,
+                p_spid,
+                p_nonce,
+                p_sig_rl,
+                sig_rl_size,
+                p_qe_report,
+                p_quote as *mut sgx_quote_t,
+                real_quote_len,
+            )
+        };
+
+        ensure!(
+            ret == sgx_status_t::SGX_SUCCESS,
+            OCallBridgeError::GetQuote(ret)
+        );
+
+        Ok((
+            qe_report,
+            Vec::from(&return_quote_buf[..real_quote_len as usize]),
+        ))
+    }
+
+    fn get_update_info(
+        &self,
+        platform_blob: sgx_platform_info_t,
+        enclave_trusted: i32,
+    ) -> OCallBridgeResult<sgx_update_info_bit_t> {
+        let mut update_info: sgx_update_info_bit_t = sgx_update_info_bit_t::default();
+
+        let result = unsafe {
+            sgx_report_attestation_status(
+                &platform_blob as *const sgx_platform_info_t,
+                enclave_trusted,
+                &mut update_info as *mut sgx_update_info_bit_t,
+            )
+        };
+
+        ensure!(
+            result == sgx_status_t::SGX_SUCCESS,
+            OCallBridgeError::GetUpdateInfo(result)
+        );
+
+        Ok(update_info)
+    }
+}
+
+fn lookup_ipv4(host: &str, port: u16) -> Result<SocketAddr, String> {
+    use std::net::ToSocketAddrs;
+
+    let addrs = (host, port)
+        .to_socket_addrs()
+        .map_err(|e| format!("{:?}", e))?;
```

Comment:
here I believe you don't need `map_err` you can directly use `?`. Not 100% sure. My reasoning:

1. you can fmt the err -> it has implemented `ToString` (so you can definetly use `map_err(|e| e.to_string())`
2. it implements `ToString` chances are very high that it also implements `Into<String>`
3. If it implements `Into<String>` you should be able to use plain `?`.

Reply:
Unfortunately that error only implements `Debug`, not `ToString` or any regular formatting.. so removing the `map_err()` will cause the compiler to complain.

Reply:
Ah ok, I somehow thought that is implements `Display`, which implicitly creates a `ToString` implementation.